iothub/device/src/IotHubBaseClient.cs

```diff
@@ -558,14 +558,24 @@ private protected static ClientPipelineBuilder BuildPipeline()
             return !isFound ? default : (T)handler;
         }
 
-        internal async Task OnMessageReceivedAsync(Message message)
+        internal async Task<MessageAcknowledgement> OnMessageReceivedAsync(Message message)
         {
             if (Logging.IsEnabled)
                 Logging.Enter(this, message, nameof(OnMessageReceivedAsync));
 
             if (message == null)
             {
-                return;
+                if (Logging.IsEnabled)
+                    Logging.Error(this, "Received a null message. Abandoning message.", nameof(OnMessageReceivedAsync));
+
+                if (_clientOptions.TransportSettings is IotHubClientMqttSettings)
+                {
+                    // MQTT does not support abandoning received messages, so we have no choice
+                    // but to complete it
+                    return MessageAcknowledgement.Complete;
+                }
+
+                return MessageAcknowledgement.Abandon;
```

Comment:
If there is no message received, what would complete or abandon mean in that context? 
I think it would be better that we check for null message upstream and not invoke this callback if the message is null. We could add a DebugAssert for our internal validation.

Reply:
This case should never happen, so I can switch this to a debug.assert, sure